packages/generator/ni_measurement_plugin_sdk_generator/client/templates/measurement_plugin_client.py.mako

```diff
@@ -58,41 +63,76 @@ class ${class_name}:
         Args:
             discovery_client: An optional discovery client.
 
+            pin_map_client: An optional pin map client.
+
             grpc_channel: An optional gRPC channel targeting a measurement service.
 
             grpc_channel_pool: An optional gRPC channel pool.
         """
-        self._initialization_lock = threading.Lock()
+        self._initialization_lock = threading.RLock()
         self._service_class = "${service_class}"
         self._grpc_channel_pool = grpc_channel_pool
         self._discovery_client = discovery_client
+        self._pin_map_client = pin_map_client
         self._stub: Optional[v2_measurement_service_pb2_grpc.MeasurementServiceStub] = None
         self._configuration_metadata = ${configuration_metadata}
         self._output_metadata = ${output_metadata}
         if grpc_channel is not None:
             self._stub = v2_measurement_service_pb2_grpc.MeasurementServiceStub(grpc_channel)
         self._create_file_descriptor()
-        
+        self._pin_map_id = ""
+        self._sites = [0]
```

Comment:
Consider replacing these with `self._pin_map_context: Optional[PinMapContext] = None`

Reply:
Replaced with `self._pin_map_context`.

Also, since sites is a repeated field, we couldn't directly assign value to sites like this `client.pin_map_context.sites = [1,2,3]`. Instead if users wants to update sites, they should be creating a new `PinMapContext` and set `sites` through dependency injection. 
For example,
```
    client = NIDCPowerSourceDCVoltageClient()
    client.pin_map_context = PinMapContext(pin_map_id="NIDCPowerSourceDCVoltageMultiSite.pinmap", sites=[0, 1])
    client.measure()
```

I've also added `sites` as an optional parameter to `register_pin_map` such that users can specify sites while registering pin map.

```
    client = NIDCPowerSourceDCVoltageClient()
    client.register_pin_map(pin_map_path, sites=[0, 1])
    client.measure()
```

Please share your thoughts on this.

Reply:
Don't use the gRPC type in public interfaces. Use the named tuple: https://github.com/ni/measurement-plugin-python/blob/main/packages/service/ni_measurement_plugin_sdk_service/session_management/_types.py#L22

Then call `_to_grpc` to convert to the gRPC type.

> Also, since sites is a repeated field, we couldn't directly assign value to sites...

The named tuple prevents assignment too, but it allows replacing fields.

```
@sites.setter
def sites(self, val):
    self._pin_map_context = self._pin_map_context._replace(sites=val)
```

> Instead if users wants to update sites, they should be creating a new PinMapContext and set sites through dependency injection.

I don't see how this is dependency injection.

> I've also added sites as an optional parameter to register_pin_map such that users can specify sites while registering pin map if needed.

I need to think about this some more.

Reply:
Thanks for the suggestion! I've replaced the gRPC type `PinMapContext` with named tuple and also, added `sites` property to set the sites externally.

Reply:
> I've also added sites as an optional parameter to register_pin_map such that users can specify sites while registering pin map if needed.

I proposed this previously because we didn't have a direct means to set sites. To set sites, we need to create new `PinMapContext` object. But now with the named tuple approach, user can set sites directly, So I've removed this sites parameter from r`egister_pin_map` which will now only update `pin_map_id` in `self._pin_map_context`.

Please share your thoughts on this.

Reply:
I think the combination of `client.register_pin_map(path)` and `client.sites = [1,2,3]` is fine.

I am not concerned with the use of a mutable list. If users do `client.pin_map_context.sites.append(4)`, it will do the right thing, because the PInMapContext named tuple is the single source of truth for the selected sites.